cabal-install/src/Distribution/Client/Dependency.hs

```diff
@@ -456,19 +456,34 @@ dontUpgradeNonUpgradeablePackages params =
         . InstalledPackageIndex.lookupPackageName
           (depResolverInstalledPkgIndex params)
 
--- NOTE: the lists of non-upgradable and non-installable packages used to be
+-- | The set of non-reinstallable packages includes those which cannot be
+-- rebuilt using a GHC installation and Hackage-published source distribution.
+-- There are a few reasons why this might be true:
+--
+--  * the package overrides its unit ID (e.g. with ghc's @-this-unit-id@ flag),
+--    which can result in multiple indistinguishable packages (having potentially
+--    different ABIs) with the same unit ID.
+--
+--  * the package contains definitions of wired-in declarations which tie
+--    it to a particular compiler (e.g. we can't build link against
+--    @base-4.18.0.0@ using GHC 9.6.1).
+--
+--  * the package does not have a complete (that is, buildable) source distribution.
+--    For instance, some packages provided by GHC rely on files outside of the
+--    source tree generated by GHC's build system.
+--
+-- Note: the list of non-upgradable/non-installable packages used to be
 -- respectively in this module and in `Distribution.Solver.Modular.Solver`.
 -- Since they were kept synced, they are now combined in the following list.
 --
--- See: https://github.com/haskell/cabal/issues/8581
+-- See: https://github.com/haskell/cabal/issues/8581 and
+-- https://github.com/haskell/cabal/issues/9064.
 nonUpgradeablePackages :: [PackageName]
 nonUpgradeablePackages =
   [ mkPackageName "base"
   , mkPackageName "ghc-bignum"
   , mkPackageName "ghc-prim"
-  , mkPackageName "ghc-boot"
   , mkPackageName "ghc"
-  , mkPackageName "ghci"
   , mkPackageName "integer-gmp"
   , mkPackageName "integer-simple"
```

Comment:
Both `integer-*` packages are reinstallable now afaik.

Reply:
@hsyl20 Is there an authoritative source about this? I can ship this change in the PR if you have a source :)

Reply:
No authoritative source, sorry.

Reply:
Even if this were true, that's only for latest ghcs, and this package set needs to span all supported ghcs, so we can't do it regardless. If we do start doing it going forward we'll need a per-ghc-version refactor of how we manage the set.